File: sdk/python/examples/framework_demo.py
"""
Example: Using A3S Code Framework with DI and Decorators

This example demonstrates the complete framework usage:
- Dependency Injection
- Decorators
- AgentFactory
- Guards, Interceptors, Pipes, Filters
"""

import asyncio

from a3s_code_framework import (
    AgentFactory,
    exception_filter,
    guard,
    injectable,
    interceptor,
    middleware,
    pipe,
)


# 1. Define services (DI)
@injectable(scope="singleton")
class LoggerService:
    """Logger service (singleton)"""

    def __init__(self, level: str = "info"):
        self.level = level

    def log(self, msg: str):
        print(f"[{self.level.upper()}] {msg}")


@injectable(scope="singleton")
class CacheService:
    """Cache service (singleton)"""

    def __init__(self):
        self._cache = {}

    def get(self, key: str):
        return self._cache.get(key)

    def set(self, key: str, value: str):
        self._cache[key] = value


@injectable(scope="scoped")
class SessionStore:
    """Session store (scoped - one per session)"""

    def __init__(self, logger: LoggerService, cache: CacheService):
        self.logger = logger
        self.cache = cache
        self.data = {}

    def save(self, key: str, value: str):
        self.logger.log(f"Saving {key}")
        self.data[key] = value
        self.cache.set(key, value)


# 2. Define middleware (bottom layer - Express-like)
@injectable()
@middleware
class LoggingMiddleware:
    """Logging middleware"""

    def __init__(self, logger: LoggerService):
        self.logger = logger

    async def handle(self, ctx):
        self.logger.log(f"Request: {ctx.get('session_id', 'unknown')}")
        return {"type": "continue"}


@injectable()
@middleware
class TimingMiddleware:
    """Timing middleware"""

    def __init__(self, logger: LoggerService):
        self.logger = logger

    async def handle(self, ctx):
        import time

        start = time.time()
        self.logger.log("⏱️  Timing started")

        # Store start time in context
        ctx["timing_start"] = start

        return {"type": "continue"}


# 3. Define guards (high-level - NestJS-like)
@injectable()
@guard
class AuthGuard:
    """Authentication guard"""

    def __init__(self, logger: LoggerService):
        self.logger = logger

    async def can_activate(self, ctx) -> bool:
        self.logger.log(f"🔐 Checking auth for session {ctx.get('session_id', 'unknown')}")
        # Mock: always allow
        return True


@injectable()
@guard
class RateLimitGuard:
    """Rate limiting guard"""

    def __init__(self, logger: LoggerService, cache: CacheService):
        self.logger = logger
        self.cache = cache
        self.max_requests = 100

    async def can_activate(self, ctx) -> bool:
        session_id = ctx.get("session_id", "unknown")
        count = int(self.cache.get(f"rate_limit:{session_id}") or 0)

        if count >= self.max_requests:
            self.logger.log(f"🚫 Rate limit exceeded for {session_id}")
            return False

        self.cache.set(f"rate_limit:{session_id}", str(count + 1))
        self.logger.log(f"✅ Rate limit check passed ({count + 1}/{self.max_requests})")
        return True


# 4. Define interceptors
@injectable()
@interceptor
class CacheInterceptor:
    """Cache interceptor"""

    def __init__(self, cache: CacheService, logger: LoggerService):
        self.cache = cache
        self.logger = logger

    async def before(self, ctx):
        session_id = ctx.get("session_id", "unknown")
        cached = self.cache.get(f"result:{session_id}")
        if cached:
            self.logger.log(f"💾 Cache hit for {session_id}")
            ctx["cached_result"] = cached

    async def after(self, ctx, result):
        session_id = ctx.get("session_id", "unknown")
        self.logger.log(f"💾 Caching result for {session_id}")
        self.cache.set(f"result:{session_id}", str(result))


# 5. Define pipes
@injectable()
@pipe
class ValidationPipe:
    """Validation pipe"""

    def __init__(self, logger: LoggerService):
        self.logger = logger

    def transform(self, value):
        self.logger.log(f"✅ Validating: {value}")
        # Mock: just return the value
        return value


# 6. Define exception filters
@injectable()
@exception_filter
class RetryFilter:
    """Retry filter"""

    def __init__(self, logger: LoggerService):
        self.logger = logger
        self.max_retries = 3

    async def catch(self, error, ctx):
        self.logger.log(f"🔄 Retry filter caught error: {error}")
        # Mock: return None (no recovery)
        return None


async def main():
    print("🚀 A3S Code Framework Example\n")

    # Create factory (NestJS-like)
    factory = AgentFactory.create("agent.hcl")

    # Register providers (DI)
    print("📦 Registering providers...")
    factory.provide(LoggerService, level="debug")
    factory.provide(CacheService)
    factory.provide(SessionStore)  # Auto-inject LoggerService + CacheService

    # Register middleware (bottom layer - Express-like)
    print("🔧 Registering middleware...")
    factory.use(LoggingMiddleware)  # Auto-inject LoggerService
    factory.use(TimingMiddleware)  # Auto-inject LoggerService

    # Register guards (high-level - NestJS-like)
    print("🛡️  Registering guards...")
    factory.use_guard(AuthGuard)  # Auto-inject LoggerService
    factory.use_guard(RateLimitGuard)  # Auto-inject LoggerService + CacheService

    # Register interceptors
    print("🔍 Registering interceptors...")
    factory.use_interceptor(CacheInterceptor)  # Auto-inject CacheService + LoggerService

    # Register pipes
    print("🔀 Registering pipes...")
    factory.use_pipe(ValidationPipe)  # Auto-inject LoggerService

    # Register filters
    print("🚨 Registering filters...")
    factory.use_filter(RetryFilter)  # Auto-inject LoggerService

    print("\n✅ All components registered with automatic dependency injection!\n")

    # Create session
    print("📝 Creating session...")
    session = factory.session("/project")

    # Run session
    print("\n🎬 Running session...\n")
    result = await session.run("List all files")

    print(f"\n📊 Result: {result}\n")
    print("🎉 Framework working correctly!")


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/node/examples/framework_demo.ts
/**
 * Example: Using A3S Code Framework with DI and Decorators (TypeScript)
 *
 * This example demonstrates the complete framework usage:
 * - Dependency Injection
 * - Decorators
 * - AgentFactory
 * - Guards, Interceptors, Pipes, Filters
 */

import 'reflect-metadata';
import {
  AgentFactory,
  Injectable,
  Scope,
  Guard,
  Interceptor,
  Pipe,
  ExceptionFilter,
  Middleware,
  MiddlewareResult,
} from '../framework';

// 1. Define services (DI)
@Injectable({ scope: Scope.SINGLETON })
class LoggerService {
  constructor(private level: string = 'info') {}

  log(msg: string): void {
    console.log(`[${this.level.toUpperCase()}] ${msg}`);
  }
}

@Injectable({ scope: Scope.SINGLETON })
class CacheService {
  private cache = new Map<string, string>();

  get(key: string): string | undefined {
    return this.cache.get(key);
  }

  set(key: string, value: string): void {
    this.cache.set(key, value);
  }
}

@Injectable({ scope: Scope.SCOPED })
class SessionStore {
  private data = new Map<string, string>();

  constructor(
    private logger: LoggerService,
    private cache: CacheService
  ) {}

  save(key: string, value: string): void {
    this.logger.log(`Saving ${key}`);
    this.data.set(key, value);
    this.cache.set(key, value);
  }
}

// 2. Define middleware (bottom layer - Express-like)
@Injectable()
@Middleware()
class LoggingMiddleware {
  constructor(private logger: LoggerService) {}

  async handle(ctx: any): Promise<{ type: string }> {
    this.logger.log(`Request: ${ctx.sessionId || 'unknown'}`);
    return { type: MiddlewareResult.CONTINUE };
  }
}

@Injectable()
@Middleware()
class TimingMiddleware {
  constructor(private logger: LoggerService) {}

  async handle(ctx: any): Promise<{ type: string }> {
    const start = Date.now();
    this.logger.log('⏱️  Timing started');
    ctx.metadata.set('timing_start', start.toString());
    return { type: MiddlewareResult.CONTINUE };
  }
}

// 3. Define guards (high-level - NestJS-like)
@Injectable()
@Guard()
class AuthGuard {
  constructor(private logger: LoggerService) {}

  async canActivate(ctx: any): Promise<boolean> {
    this.logger.log(`🔐 Checking auth for session ${ctx.sessionId || 'unknown'}`);
    return true; // Mock: always allow
  }
}

@Injectable()
@Guard()
class RateLimitGuard {
  private maxRequests = 100;

  constructor(
    private logger: LoggerService,
    private cache: CacheService
  ) {}

  async canActivate(ctx: any): Promise<boolean> {
    const sessionId = ctx.sessionId || 'unknown';
    const count = parseInt(this.cache.get(`rate_limit:${sessionId}`) || '0');

    if (count >= this.maxRequests) {
      this.logger.log(`🚫 Rate limit exceeded for ${sessionId}`);
      return false;
    }

    this.cache.set(`rate_limit:${sessionId}`, (count + 1).toString());
    this.logger.log(`✅ Rate limit check passed (${count + 1}/${this.maxRequests})`);
    return true;
  }
}

// 4. Define interceptors
@Injectable()
@Interceptor()
class CacheInterceptor {
  constructor(
    private cache: CacheService,
    private logger: LoggerService
  ) {}

  async before(ctx: any): Promise<void> {
    const sessionId = ctx.sessionId || 'unknown';
    const cached = this.cache.get(`result:${sessionId}`);
    if (cached) {
      this.logger.log(`💾 Cache hit for ${sessionId}`);
      ctx.cachedResult = cached;
    }
  }

  async after(ctx: any, result: any): Promise<void> {
    const sessionId = ctx.sessionId || 'unknown';
    this.logger.log(`💾 Caching result for ${sessionId}`);
    this.cache.set(`result:${sessionId}`, JSON.stringify(result));
  }
}

// 5. Define pipes
@Injectable()
@Pipe()
class ValidationPipe {
  constructor(private logger: LoggerService) {}

  transform(value: any): any {
    this.logger.log(`✅ Validating: ${JSON.stringify(value)}`);
    return value; // Mock: just return the value
  }
}

// 6. Define exception filters
@Injectable()
@ExceptionFilter()
class RetryFilter {
  private maxRetries = 3;

  constructor(private logger: LoggerService) {}

  async catch(error: Error, ctx: any): Promise<any> {
    this.logger.log(`🔄 Retry filter caught error: ${error.message}`);
    return null; // Mock: no recovery
  }
}

async function main() {
  console.log('🚀 A3S Code Framework Example (TypeScript)\n');

  // Create factory (NestJS-like)
  const factory = AgentFactory.create('agent.hcl');

  // Register providers (DI)
  console.log('📦 Registering providers...');
  factory.provide(LoggerService, 'debug');
  factory.provide(CacheService);
  factory.provide(SessionStore); // Auto-inject LoggerService + CacheService

  // Register middleware (bottom layer - Express-like)
  console.log('🔧 Registering middleware...');
  factory.use(LoggingMiddleware); // Auto-inject LoggerService
  factory.use(TimingMiddleware); // Auto-inject LoggerService

  // Register guards (high-level - NestJS-like)
  console.log('🛡️  Registering guards...');
  factory.useGuard(AuthGuard); // Auto-inject LoggerService
  factory.useGuard(RateLimitGuard); // Auto-inject LoggerService + CacheService

  // Register interceptors
  console.log('🔍 Registering interceptors...');
  factory.useInterceptor(CacheInterceptor); // Auto-inject CacheService + LoggerService

  // Register pipes
  console.log('🔀 Registering pipes...');
  factory.usePipe(ValidationPipe); // Auto-inject LoggerService

  // Register filters
  console.log('🚨 Registering filters...');
  factory.useFilter(RetryFilter); // Auto-inject LoggerService

  console.log('\n✅ All components registered with automatic dependency injection!\n');

  // Create session
  console.log('📝 Creating session...');
  const session = factory.session('/project');

  // Run session
  console.log('\n🎬 Running session...\n');
  const result = await session.run('List all files');

  console.log(`\n📊 Result:`, result);
  console.log('\n🎉 Framework working correctly!');
}

main().catch(console.error);
